From ef7261816d6126d1e1ab30d7f7ddbfb6114e5ddb Mon Sep 17 00:00:00 2001
From: Nic Costa <nic.costa@gmail.com>
Date: Thu, 11 Mar 2021 13:01:03 -0600
Subject: [PATCH 13/14] add snap version /3/0/4001

Note we query the snap version through the SNAPD REST API rather
than calling "snap info <project-name>" because the latter fails
with a timeout and an error in the journal with a seemingly
unrelated error message:

AVC apparmor="DENIED" operation="open"
profile="snap.pelion-edge.edge-core" name="/home/admin/.snap/auth.json"
pid=16456 comm="snap" requested_mask="r" denied_mask="r" fsuid=0
ouid=1000
---
 edge-client/gateway_stats.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/edge-client/gateway_stats.c b/edge-client/gateway_stats.c
index 2477443..b6aab2e 100644
--- a/edge-client/gateway_stats.c
+++ b/edge-client/gateway_stats.c
@@ -36,6 +36,7 @@
 #define GATEWAY_STATS_RAM_TOTAL_RES_ID 3322
 #define GATEWAY_STATS_DISK_FREE_RES_ID 3323
 #define GATEWAY_STATS_DISK_TOTAL_RES_ID 3324
+#define GATEWAY_STATS_VERSION_RES_ID 4001
 
 /**
  * \struct cpu_info
@@ -247,6 +248,23 @@ void gsr_add_gateway_stats_resources()
     int64_t int_default = 0;
     float float_default = 0;
     int64_t int_actual;
+    char version[64];
+
+    // snap version
+    memset(version, 0, sizeof(version));
+    const char cmd_version[] = "curl -sS -H \"Content-Type: application/json\" --unix-socket /run/snapd.socket http://localhost/v2/snaps/" SNAPCRAFT_PROJECT_NAME " | jq -r .result.version";
+    if (sys_exec(cmd_version, version, sizeof(version)) != 0) {
+        strcpy(version, "-1");
+    }
+    gsr_create_resource(GATEWAY_STATS_OBJ_ID,
+                        0,
+                        GATEWAY_STATS_VERSION_RES_ID,
+                        "edge snap version",
+                        LWM2M_STRING,
+                        OPERATION_READ,
+                        (uint8_t *)version,
+                        strlen(version),
+                        NULL);
 
     // cpu temp
     gsr_create_resource(GATEWAY_STATS_OBJ_ID,
-- 
2.30.1 (Apple Git-130)

